ext/MitsubaLoader/CMitsubaMaterialCompilerFrontend.cpp

```diff
@@ -0,0 +1,597 @@
+                dst.source = IR::INode::EPS_CONSTANT;
+                dst.value.constant = IR::INode::color_t(1.f, 0.f, 0.f);//red in case when didnt find texture
+            }
+        }
+        else
+        {
+            dst.source = IR::INode::EPS_CONSTANT;
+            dst.value.constant = src.value.vvalue;
+        }
+    };
+
+    constexpr IR::CMicrofacetSpecularBSDFNode::E_NDF ndfMap[4]{
+        IR::CMicrofacetSpecularBSDFNode::ENDF_BECKMANN,
+        IR::CMicrofacetSpecularBSDFNode::ENDF_GGX,
+        IR::CMicrofacetSpecularBSDFNode::ENDF_PHONG,
+        IR::CMicrofacetSpecularBSDFNode::ENDF_ASHIKHMIN_SHIRLEY
+    };
+
+    IR::INode* root = ir->allocRootNode<IR::CMaterialNode>();
+    root->children.count = 1u;
+
+    bool twosided = false;
+    IR::INode::SParameter<IR::INode::color_t> opacity;
+    {
+        opacity.source = IR::INode::EPS_CONSTANT;
+        opacity.value.constant = IR::INode::color_t(1.f);
+    }
+    bool thin = false;
+
+    const CElementBSDF* current = _bsdf;
+
+    core::queue<const CElementBSDF*> bsdfQ;
+    bsdfQ.push(_bsdf);
+    core::queue<IR::INode*> nodeQ;
+    //nodeQ.push(root);
+    uint32_t childrenCountdown = 1u;
+    IR::INode* parent = root;//nodeQ.front();
+
+    while (!bsdfQ.empty())
+    {
+        current = bsdfQ.front();
+        bsdfQ.pop();
+
+        IR::INode* nextSym;
+        const auto currType = current->type;
+        switch (currType)
+        {
+        case CElementBSDF::TWO_SIDED:
+            twosided = true;
+            bsdfQ.push(current->twosided.bsdf[0]);
+            continue;
+            break;
+        case CElementBSDF::MASK:
+            getSpectrumOrTexture(current->mask.opacity, opacity);
+            bsdfQ.push(current->mask.bsdf[0]);
+            continue;
+            break;
+        case CElementBSDF::DIFFUSE:
+        case CElementBSDF::ROUGHDIFFUSE:
+            nextSym = ir->allocNode<IR::CMicrofacetDiffuseBSDFNode>();
+            getSpectrumOrTexture(current->diffuse.reflectance, static_cast<IR::CMicrofacetDiffuseBSDFNode*>(nextSym)->reflectance);
+            if (currType==CElementBSDF::ROUGHDIFFUSE)
+            {
+                getFloatOrTexture(current->diffuse.alpha, static_cast<IR::CMicrofacetDiffuseBSDFNode*>(nextSym)->alpha_u);
+                getFloatOrTexture(current->diffuse.alpha, static_cast<IR::CMicrofacetDiffuseBSDFNode*>(nextSym)->alpha_v);
+            }
+            else
+            {
+                static_cast<IR::CMicrofacetDiffuseBSDFNode*>(nextSym)->setSmooth();
+            }
+            break;
+        case CElementBSDF::CONDUCTOR:
+        case CElementBSDF::ROUGHCONDUCTOR:
+        {
+            nextSym = ir->allocNode<IR::CMicrofacetSpecularBSDFNode>();
+            auto* node = static_cast<IR::CMicrofacetSpecularBSDFNode*>(nextSym);
+            node->shadowing = IR::CMicrofacetSpecularBSDFNode::EST_SMITH;
+            node->scatteringMode = IR::CMicrofacetSpecularBSDFNode::ESM_REFLECT;
```

Comment:
what's scattering mode?

Reply:
reflect or transmit, stolen from nvidia MDL
For example dielectric in IR is encoded as fresnel-weighted reflection/transmission